File: app/src/main/java/maderski/bluetoothautoplaymusic/bluetoothactions/BTDisconnectActions.java
package maderski.bluetoothautoplaymusic.bluetoothactions;

import android.content.Context;
import android.media.AudioManager;
import android.os.Handler;
import android.util.Log;

import maderski.bluetoothautoplaymusic.controls.PlayMusicControl;
import maderski.bluetoothautoplaymusic.controls.RingerControl;
import maderski.bluetoothautoplaymusic.controls.VolumeControl;
import maderski.bluetoothautoplaymusic.controls.WifiControl;
import maderski.bluetoothautoplaymusic.helpers.TimeHelper;
import maderski.bluetoothautoplaymusic.helpers.LaunchAppHelper;
import maderski.bluetoothautoplaymusic.BAPMNotification;
import maderski.bluetoothautoplaymusic.PackageTools;
import maderski.bluetoothautoplaymusic.services.BTDisconnectService;
import maderski.bluetoothautoplaymusic.services.WakeLockService;
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMDataPreferences;
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMPreferences;
import maderski.bluetoothautoplaymusic.utils.ServiceUtils;

/**
 * Created by Jason on 6/3/17.
 */

public class BTDisconnectActions {
    private static final String TAG = "BTDisconnectActions";

    private final Context context;
    private final BAPMNotification mBAPMNotification;
    private final VolumeControl mVolumeControl;
    private final PlayMusicControl mPlayMusicControl;

    public BTDisconnectActions(Context context){
        this.context = context;
        this.mBAPMNotification = new BAPMNotification();
        this.mVolumeControl = new VolumeControl(context);
        this.mPlayMusicControl = new PlayMusicControl(context);
    }

    //Removes mNotification and if set releases wakelock, puts the ringer back to normal,
    //pauses the music
    public void actionsOnBTDisconnect(){
        LaunchAppHelper launchAppHelper = new LaunchAppHelper();
        RingerControl ringerControl = new RingerControl(context);

        removeBAPMNotification();
        pauseMusic();
        turnOffPriorityMode(ringerControl);
        sendAppToBackground(launchAppHelper);
        closeWaze(launchAppHelper);
        setWifiOn(launchAppHelper);
        stopKeepingScreenOn();

        setVolumeBack(ringerControl);

        stopService();
    }

    private void stopService() {
        BAPMDataPreferences.INSTANCE.setRanActionsOnBtConnect(context, false);
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                ServiceUtils.INSTANCE.stopService(context, BTDisconnectService.class, BTDisconnectService.TAG);
            }
        }, 5000);
    }

    private void removeBAPMNotification(){
        boolean canShowNotification = BAPMPreferences.INSTANCE.getShowNotification(context);

        if(canShowNotification) {
            mBAPMNotification.removeBAPMMessage(context);
        }
    }

    private void pauseMusic(){
        boolean playMusic = BAPMPreferences.INSTANCE.getAutoPlayMusic(context);
        if (playMusic) {
            mPlayMusicControl.pause();
        }
    }

    private void sendAppToBackground(LaunchAppHelper launchAppHelper){
        boolean sendToBackground = BAPMPreferences.INSTANCE.getSendToBackground(context);
        if (sendToBackground) {
            launchAppHelper.sendEverythingToBackground(context);
        }
    }

    private void turnOffPriorityMode(RingerControl ringerControl){

        boolean priorityMode = BAPMPreferences.INSTANCE.getPriorityMode(context);
        if (priorityMode) {
            int currentRinger = BAPMDataPreferences.INSTANCE.getCurrentRingerSet(context);
            try {
                switch (currentRinger) {
                    case AudioManager.RINGER_MODE_SILENT:
                        Log.d(TAG, "Phone is on Silent");
                        break;
                    case AudioManager.RINGER_MODE_VIBRATE:
                        ringerControl.vibrateOnly();
                        break;
                    case AudioManager.RINGER_MODE_NORMAL:
                        ringerControl.soundsON();
                        break;
                }
            } catch (Exception e) {
                Log.e(TAG, e.getMessage());
            }
        }
    }

    private void closeWaze(LaunchAppHelper launchAppHelper){
        boolean closeWaze = BAPMPreferences.INSTANCE.getCloseWazeOnDisconnect(context)
                && launchAppHelper.checkPkgOnPhone(context, PackageTools.PackageName.WAZE)
                && BAPMPreferences.INSTANCE.getMapsChoice(context).equals(PackageTools.PackageName.WAZE);
        if(closeWaze) {
            launchAppHelper.closeWazeOnDisconnect(context);
        }
    }

    private void setWifiOn(LaunchAppHelper launchAppHelper){
        boolean isWifiOffDevice = BAPMDataPreferences.INSTANCE.getIsTurnOffWifiDevice(context);
        if(isWifiOffDevice){
            int eveningStartTime = BAPMPreferences.INSTANCE.getEveningStartTime(context);
            int eveningEndTime = BAPMPreferences.INSTANCE.getEveningEndTime(context);

            int current24hrTime = TimeHelper.getCurrent24hrTime();

            TimeHelper timeHelperEvening = new TimeHelper(eveningStartTime, eveningEndTime, current24hrTime);
            boolean canLaunch = timeHelperEvening.isWithinTimeSpan();
            String directionLocation = canLaunch ? LaunchAppHelper.DirectionLocations.HOME : LaunchAppHelper.DirectionLocations.WORK;

            boolean canChangeWifiState = !BAPMPreferences.INSTANCE.getWifiUseMapTimeSpans(context)
                    || (canLaunch && launchAppHelper.canLaunchOnThisDay(context, directionLocation));
            if(canChangeWifiState && !WifiControl.INSTANCE.isWifiON(context)) {
                WifiControl.INSTANCE.wifiON(context, true);
            }
            BAPMDataPreferences.INSTANCE.setIsTurnOffWifiDevice(context, false);
        }
    }

    private void stopKeepingScreenOn(){
        boolean screenON = BAPMPreferences.INSTANCE.getKeepScreenON(context);
        if (screenON) {
            ServiceUtils.INSTANCE.stopService(context, WakeLockService.class, WakeLockService.TAG);
        }
    }

    private void setVolumeBack(RingerControl ringerControl){
        boolean volumeMAX = BAPMPreferences.INSTANCE.getMaxVolume(context);
        boolean setOriginalVolume = BAPMPreferences.INSTANCE.getRestoreNotificationVolume(context);

        if (volumeMAX && setOriginalVolume) {
            mVolumeControl.setToOriginalVolume(ringerControl);
        }
    }
}

File: app/src/main/java/maderski/bluetoothautoplaymusic/bluetoothactions/BTDisconnectActions.kt
package maderski.bluetoothautoplaymusic.bluetoothactions

import android.content.Context
import android.media.AudioManager
import android.os.Handler
import android.util.Log
import maderski.bluetoothautoplaymusic.controls.PlayMusicControl
import maderski.bluetoothautoplaymusic.controls.RingerControl
import maderski.bluetoothautoplaymusic.controls.VolumeControl
import maderski.bluetoothautoplaymusic.controls.WifiControl
import maderski.bluetoothautoplaymusic.helpers.LaunchAppHelper
import maderski.bluetoothautoplaymusic.helpers.LaunchAppHelper.DirectionLocation
import maderski.bluetoothautoplaymusic.helpers.PackageHelper.MapApps.WAZE
import maderski.bluetoothautoplaymusic.helpers.TimeHelper
import maderski.bluetoothautoplaymusic.notification.BAPMNotification
import maderski.bluetoothautoplaymusic.services.BTDisconnectService
import maderski.bluetoothautoplaymusic.services.WakeLockService
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMDataPreferences
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMPreferences
import maderski.bluetoothautoplaymusic.utils.ServiceUtils
import org.koin.core.KoinComponent
import org.koin.core.inject

/**
 * Created by Jason on 6/3/17.
 */

class BTDisconnectActions(private val context: Context): KoinComponent {
    private val preferences: BAPMPreferences by inject()
    private val dataPreferences: BAPMDataPreferences by inject()

    private val mBAPMNotification: BAPMNotification = BAPMNotification(context)
    private val mVolumeControl: VolumeControl = VolumeControl(context)
    private val mPlayMusicControl: PlayMusicControl = PlayMusicControl(context)

    //Removes mNotification and if set releases wakelock, puts the ringer back to normal,
    //pauses the music
    fun actionsOnBTDisconnect() {
        val launchAppHelper = LaunchAppHelper(context)
        val ringerControl = RingerControl(context)

        removeBAPMNotification()
        pauseMusic()
        turnOffPriorityMode(ringerControl)
        sendAppToBackground(launchAppHelper)
        closeWaze(launchAppHelper)
        setWifiOn(launchAppHelper)
        stopKeepingScreenOn()

        setVolumeBack(ringerControl)

        stopService()
    }

    private fun stopService() {
        dataPreferences.setRanActionsOnBtConnect(false)
        Handler().postDelayed({ ServiceUtils.stopService(context, BTDisconnectService::class.java, BTDisconnectService.TAG) }, 5000)
    }

    private fun removeBAPMNotification() {
        val canShowNotification = preferences.getShowNotification()

        if (canShowNotification) {
            mBAPMNotification.removeBAPMMessage()
        }
    }

    private fun pauseMusic() {
        val playMusic = preferences.getAutoPlayMusic()
        if (playMusic) {
            mPlayMusicControl.pause()
        }
    }

    private fun sendAppToBackground(launchAppHelper: LaunchAppHelper) {
        val sendToBackground = preferences.getSendToBackground()
        if (sendToBackground) {
            launchAppHelper.sendEverythingToBackground(context)
        }
    }

    private fun turnOffPriorityMode(ringerControl: RingerControl) {

        val priorityMode = preferences.getPriorityMode()
        if (priorityMode) {
            val currentRinger = dataPreferences.getCurrentRingerSet()
            try {
                when (currentRinger) {
                    AudioManager.RINGER_MODE_SILENT -> Log.d(TAG, "Phone is on Silent")
                    AudioManager.RINGER_MODE_VIBRATE -> ringerControl.vibrateOnly()
                    AudioManager.RINGER_MODE_NORMAL -> ringerControl.soundsON()
                }
            } catch (e: Exception) {
                Log.e(TAG, e.message)
            }

        }
    }

    private fun closeWaze(launchAppHelper: LaunchAppHelper) {
        val closeWaze = (preferences.getCloseWazeOnDisconnect()
                && launchAppHelper.isAbleToLaunch(WAZE.packageName)
                && preferences.getMapsChoice() == WAZE.packageName)
        if (closeWaze) {
            launchAppHelper.closeWazeOnDisconnect()
        }
    }

    private fun setWifiOn(launchAppHelper: LaunchAppHelper) {
        val isWifiOffDevice = dataPreferences.getIsTurnOffWifiDevice()
        if (isWifiOffDevice) {
            val eveningStartTime = preferences.getEveningStartTime()
            val eveningEndTime = preferences.getEveningEndTime()

            val current24hrTime = TimeHelper.current24hrTime

            val timeHelperEvening = TimeHelper(eveningStartTime, eveningEndTime, current24hrTime)
            val canLaunch = timeHelperEvening.isWithinTimeSpan
            val directionLocation = if (canLaunch) DirectionLocation.HOME else DirectionLocation.WORK

            val canChangeWifiState = preferences.getWifiUseMapTimeSpans().not() || canLaunch && launchAppHelper.canLaunchOnThisDay(directionLocation)
            if (canChangeWifiState && !WifiControl.isWifiON(context)) {
                WifiControl.wifiON(context, true)
            }
            dataPreferences.setIsTurnOffWifiDevice(false)
        }
    }

    private fun stopKeepingScreenOn() {
        val screenON = preferences.getKeepScreenON()
        if (screenON) {
            ServiceUtils.stopService(context, WakeLockService::class.java, WakeLockService.TAG)
        }
    }

    private fun setVolumeBack(ringerControl: RingerControl) {
        val volumeMAX = preferences.getMaxVolume()
        val setOriginalVolume = preferences.getRestoreNotificationVolume()

        if (volumeMAX && setOriginalVolume) {
            mVolumeControl.setToOriginalVolume(ringerControl)
        }
    }

    companion object {
        private const val TAG = "BTDisconnectActions"
    }
}
